File: app/src/main/java/com/canli/oya/traininventoryroom/adapters/TrainAdapter.java
package com.canli.oya.traininventoryroom.adapters;

import android.databinding.DataBindingUtil;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.canli.oya.traininventoryroom.R;
import com.canli.oya.traininventoryroom.data.entities.TrainEntry;
import com.canli.oya.traininventoryroom.databinding.TrainItemBinding;

import java.util.List;

public class TrainAdapter extends RecyclerView.Adapter<TrainAdapter.TrainViewHolder>{

    private List<TrainEntry> mTrainList;
    private final TrainItemClickListener mClickListener;

    public TrainAdapter(TrainItemClickListener listener) {
        mClickListener = listener;
    }

    @NonNull
    @Override
    public TrainViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        TrainItemBinding binding = DataBindingUtil
                .inflate(LayoutInflater.from(parent.getContext()), R.layout.train_item,
                        parent, false);
        binding.setTrainItemClick(mClickListener);
        return new TrainViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull TrainViewHolder holder, int position) {
        TrainEntry currentTrain = mTrainList.get(position);
        holder.binding.setTrain(currentTrain);
        holder.binding.executePendingBindings();
    }

    @Override
    public int getItemCount() {
        return mTrainList == null ? 0 : mTrainList.size();
    }

    public void setTrains(List<TrainEntry> newList){
        mTrainList = newList;
        notifyDataSetChanged();
    }

    class TrainViewHolder extends RecyclerView.ViewHolder{

        final TrainItemBinding binding;

        TrainViewHolder(TrainItemBinding binding){
            super(binding.getRoot());
            this.binding = binding;
        }
    }

    public interface TrainItemClickListener {
        void onListItemClick(int trainId);
    }
}

File: app/src/main/java/com/canli/oya/traininventoryroom/adapters/TrainAdapter.kt
package com.canli.oya.traininventoryroom.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.canli.oya.traininventoryroom.R
import com.canli.oya.traininventoryroom.data.TrainEntry
import com.canli.oya.traininventoryroom.databinding.ItemTrainBinding

class TrainAdapter (private val mClickListener: TrainItemClickListener) : RecyclerView.Adapter<TrainAdapter.TrainViewHolder>() {

    var trainList: List<TrainEntry>? = null
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TrainViewHolder {
        val binding = DataBindingUtil
                .inflate<ItemTrainBinding>(LayoutInflater.from(parent.context), R.layout.item_train,
                        parent, false)
        binding.trainItemClick = mClickListener
        return TrainViewHolder(binding)
    }

    override fun onBindViewHolder(holder: TrainViewHolder, position: Int) {
        val currentTrain = trainList?.get(position)
        holder.binding.train = currentTrain
        holder.binding.executePendingBindings()
    }

    override fun getItemCount() = trainList?.size ?: 0

    inner class TrainViewHolder(val binding: ItemTrainBinding) : RecyclerView.ViewHolder(binding.root)

    interface TrainItemClickListener {
        fun onListItemClick(trainId: Int)
    }
}
